### DISCUSSION TITLE
Provider-push : How to specify a parameter ?

### DISCUSSION THREAD

**Discussion by fbattello** [eclipse-edc/Connector#2983] [Q&A] [answered]
Hello everybody,

I'm a beginner with the EDC trying to achieve a scenario based on the [provider-push](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-07-provider-push-http) sample.
I'd like the API exposed by the provider ecosystem to return different data depending on a parameter sent by the consumer.

I've reproduced the [provider-push](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-07-provider-push-http) sample, replacing the placeholder API with my own API.
Now I would like to extend the scenario to take into account an additional parameter.
But I have no clue. Any help would be greatly accepted.

Pretty logically it's unfortunately too late - once in the [ReceiverHandler ](https://github.com/eclipse-edc/Samples/blob/main/transfer/transfer-07-provider-push-http/provider-push-http-backend-service/src/main/java/org/eclipse/edc/handler/ReceiverHandler.java)dealing with an [HTTPExchange ](https://docs.oracle.com/javase/8/docs/jre/api/net/httpserver/spec/com/sun/net/httpserver/HttpExchange.html)object - to retrieve any property.

I guess a solution would imply to inherit from some base class and implement custom behavior.
I had a look at this [discussion ](https://github.com/eclipse-edc/Connector/discussions/1918) that seems to be related, but don't have enough inner understanding to find a way to achieve the whole thing.

Let's say we want to retrieve [user#1](https://jsonplaceholder.typicode.com/users/1) from [https://jsonplaceholder.typicode.com/users](https://jsonplaceholder.typicode.com/users) by specifying a ``userId`` property in the curl command - [Step #8](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-07-provider-push-http#8-start-the-transfer).
How should we modify the code in order to retrieve the extra-parameter on the provider-side and build a different response ?

Thanks for your help.

**Comment by ndr-brt** [comment]
You can implement a [`HttpParamsDecorator`](https://github.com/eclipse-edc/Connector/blob/ba5b3369d9883da9dcc66480fe098c7332cd44f2/spi/data-plane/data-plane-http-spi/src/main/java/org/eclipse/edc/connector/dataplane/http/spi/HttpParamsDecorator.java#L24) and register it to the [`HttpRequestParamsProvider`](https://github.com/eclipse-edc/Connector/blob/ba5b3369d9883da9dcc66480fe098c7332cd44f2/spi/data-plane/data-plane-http-spi/src/main/java/org/eclipse/edc/connector/dataplane/http/spi/HttpRequestParamsProvider.java#L23).

**Reply by fbattello** [reply]
Thanks !
Following your advice please let me know if I'm on the right path.
According to the data-plane-http extension [documentation](https://github.com/eclipse-edc/Connector/tree/main/extensions/data-plane/data-plane-http) I can create a custom extension that will register my custom ``HttpParamsDecorator``.
In order for the connector to load my extension, do I have to write my custom runtime ?
Something like the code found in this [Sample](https://github.com/eclipse-edc/Samples/tree/main/other/custom-runtime) ?

**Reply by ndr-brt** [reply, marked as answer]
you don't need a custom runtime, you just need to write an extension.
you could take a look [at this sample](https://github.com/eclipse-edc/Samples/tree/main/transfer/transfer-01-file-transfer), the `transfer-file-local` is a module that provides an extension.

**Reply by fbattello** [reply]
Ok. Think I've got it thanks to this last sample.
